Source/CarthageKit/Project.swift

```diff
@@ -398,20 +398,33 @@ public final class Project {
 	}
 
 	/// Loads the dependencies for the given dependency, at the given version.
-	private func dependencies(for dependency: Dependency<PinnedVersion>) -> SignalProducer<Dependency<VersionSpecifier>, CarthageError> {
+	private func dependencies(for dependency: Dependency<PinnedVersion>, usingCheckout: Bool) -> SignalProducer<Dependency<VersionSpecifier>, CarthageError> {
 
 		switch dependency.project {
 		case .git, .gitHub:
-			let revision = dependency.version.commitish
-			return self.cloneOrFetchDependency(dependency.project, commitish: revision)
-				.flatMap(.concat) { repositoryURL in
-					return contentsOfFileInRepository(repositoryURL, CarthageProjectCartfilePath, revision: revision)
-				}
-				.flatMapError { _ in .empty }
-				.attemptMap(Cartfile.from(string:))
+			let cartfileProducer: SignalProducer<Cartfile, CarthageError>
+			let dependencyCheckoutURL = self.directoryURL.appendingPathComponent(dependency.project.relativePath)
+			 // If the checkout directory doesn't exist, it may be a binary download, so fall back to fetching the cartfile
+			if usingCheckout && FileManager.default.fileExists(atPath: dependencyCheckoutURL.path) {
```

Comment:
With https://github.com/BobElDevil/Carthage/commit/c9d6c6fb059206590b1cf1ba0d532e2deae7d3af, there‘s a race condition between this `fileExists` call and when the SignalProducer is started into the file being read.

To minimize this time difference, we‘d want the `fileExists` call as part of the SignalProducer chain.